File: tools/ringlab/ringlab/commands.py
"""Catalogo de comandos y constructores de payload.

OJO: todo lo de aqui procede de la documentacion del R11M (narey83/vitals-smart-ring-app)
y esta SIN VERIFICAR contra el R21M. Cada entrada lleva su estado; el objetivo de la
Etapa 1 es ir confirmando o corrigiendo estos valores y volcarlos a docs/PROTOCOLO-R21M.md.

Ademas las dos fuentes del proyecto de referencia se contradicen en el grupo de las
consultas de historial: PROTOCOL.md las documenta en el grupo 05 y COMMANDS.md lista
GetHistoryHeart como 0x0206. Se incluyen ambas variantes para poder probarlas.
"""

from __future__ import annotations

import datetime as _dt
from enum import IntEnum

from .protocol import build_frame

# --- Comandos app -> anillo -------------------------------------------------

SET_TIME = 0x0100
SETTING_HEART_MONITOR = 0x010C
SETTING_SPO2_MONITOR = 0x0126

GET_DEVICE_INFO = 0x0200
GET_DEVICE_MAC = 0x0202
GET_DEVICE_NAME = 0x0203
GET_NOW_STEP = 0x020C
GET_ALL_REAL_DATA = 0x0220

# Variante "grupo 02" de las consultas de historial (segun COMMANDS.md)
GET_HISTORY_HEART_V2 = 0x0206
GET_HISTORY_BLOOD_V2 = 0x0208

START_MEASUREMENT = 0x032F

# Variante "grupo 05" de las consultas de historial (segun PROTOCOL.md)
HISTORY_SPORT = 0x0502
HISTORY_SLEEP = 0x0504
HISTORY_HEART = 0x0506
HISTORY_BLOOD = 0x0508
HISTORY_SPO2 = 0x0509

# --- Notificaciones anillo -> app -------------------------------------------

EVT_MEASURE_DONE = 0x040E

EVT_STORED_HEART = 0x0515
EVT_STORED_BLOOD = 0x0517
EVT_STORED_SPO2 = 0x0518
EVT_STORED_SLEEP = 0x0513

EVT_LIVE_HEART = 0x0601
EVT_LIVE_SPO2 = 0x0602
EVT_LIVE_BLOOD = 0x0603
EVT_LIVE_PPG = 0x0604
EVT_LIVE_ECG = 0x0605

NAMES: dict[int, str] = {
    SET_TIME: "SettingTime",
    SETTING_HEART_MONITOR: "SettingHeartMonitor",
    SETTING_SPO2_MONITOR: "SettingSpo2Monitor",
    GET_DEVICE_INFO: "GetDeviceInfo",
    GET_DEVICE_MAC: "GetDeviceMac",
    GET_DEVICE_NAME: "GetDeviceName",
    GET_NOW_STEP: "GetNowStep",
    GET_ALL_REAL_DATA: "GetAllRealDataFromDevice",
    GET_HISTORY_HEART_V2: "GetHistoryHeart(v2)",
    GET_HISTORY_BLOOD_V2: "GetHistoryBlood(v2)",
    START_MEASUREMENT: "AppStartMeasurement",
    HISTORY_SPORT: "Health_HistorySport",
    HISTORY_SLEEP: "Health_HistorySleep",
    HISTORY_HEART: "Health_HistoryHeart",
    HISTORY_BLOOD: "Health_HistoryBlood",
    HISTORY_SPO2: "Health_HistorySpo2",
    EVT_MEASURE_DONE: "MeasureComplete",
    EVT_STORED_SLEEP: "StoredSleepChunk",
    EVT_STORED_HEART: "StoredHeartRecord",
    EVT_STORED_BLOOD: "StoredBloodRecord",
    EVT_STORED_SPO2: "StoredSpo2Record",
    EVT_LIVE_HEART: "Real_UploadHeart",
    EVT_LIVE_SPO2: "Real_UploadBloodOxygen",
    EVT_LIVE_BLOOD: "Real_UploadBlood",
    EVT_LIVE_PPG: "Real_UploadPPG",
    EVT_LIVE_ECG: "Real_UploadECG",
}


def name_of(code: int) -> str:
    return NAMES.get(code, f"desconocido_{code:04X}")


class MeasureType(IntEnum):
    HEART = 0x00
    BLOOD_PRESSURE = 0x01
    SPO2 = 0x02


def frame(code: int, payload: bytes = b"") -> bytes:
    return build_frame(code >> 8, code & 0xFF, payload)


# --- Constructores ----------------------------------------------------------


def set_time(when: _dt.datetime) -> bytes:
    """01 00 con [anio uint16 LE][mes][dia][hora][min][seg][00]."""
    payload = (
        when.year.to_bytes(2, "little")
        + bytes([when.month, when.day, when.hour, when.minute, when.second, 0x00])
    )
    return frame(SET_TIME, payload)


def get_device_info() -> bytes:
    """02 00 con payload ASCII 'GC'."""
    return frame(GET_DEVICE_INFO, b"GC")


def get_now_step() -> bytes:
    return frame(GET_NOW_STEP)


def start_measurement(kind: MeasureType) -> bytes:
    return frame(START_MEASUREMENT, bytes([0x01, int(kind)]))


def stop_measurement() -> bytes:
    return frame(START_MEASUREMENT, bytes([0x00, 0x00]))


def set_heart_monitor(enabled: bool, minutes: int) -> bytes:
    return frame(SETTING_HEART_MONITOR, bytes([1 if enabled else 0, minutes & 0xFF]))


def set_spo2_monitor(enabled: bool, minutes: int) -> bytes:
    return frame(SETTING_SPO2_MONITOR, bytes([1 if enabled else 0, minutes & 0xFF]))


HISTORY_QUERIES: dict[str, int] = {
    "hr": HISTORY_HEART,
    "bp": HISTORY_BLOOD,
    "spo2": HISTORY_SPO2,
    "sleep": HISTORY_SLEEP,
    "sport": HISTORY_SPORT,
}


def query_history(kind: str) -> bytes:
    try:
        return frame(HISTORY_QUERIES[kind])
    except KeyError:
        raise ValueError(
            f"historial desconocido {kind!r}; opciones: {', '.join(HISTORY_QUERIES)}"
        ) from None

File: app/lib/protocol/commands.dart
import 'dart:typed_data';

import 'frame.dart';

/// Catalogo de comandos. Puerto de `tools/ringlab/ringlab/commands.py`.
///
/// El estado de verificacion de cada comando esta en `docs/PROTOCOLO-R21M.md`.

// --- App -> anillo ---------------------------------------------------------

const int cmdSetTime = 0x0100; // VERIFICADO
const int cmdSettingHeartMonitor = 0x010C;
const int cmdSettingSpo2Monitor = 0x0126;

const int cmdGetDeviceInfo = 0x0200; // VERIFICADO
const int cmdGetNowStep = 0x020C; // VERIFICADO (responde; formato sin resolver)

const int cmdStartMeasurement = 0x032F; // VERIFICADO los 3 tipos

const int cmdHistorySport = 0x0502;
const int cmdHistorySleep = 0x0504; // VERIFICADO que es aceptado
const int cmdHistoryHeart = 0x0506; // VERIFICADO que es aceptado
const int cmdHistoryBlood = 0x0508;
const int cmdHistorySpo2 = 0x0509;

// --- Anillo -> app ---------------------------------------------------------

const int evtMeasureDone = 0x040E; // VERIFICADO

const int evtStoredSleep = 0x0513;
const int evtStoredHeart = 0x0515;
const int evtStoredBlood = 0x0517;
const int evtStoredSpo2 = 0x0518;

const int evtLiveHeart = 0x0601; // VERIFICADO
const int evtLiveSpo2 = 0x0602; // VERIFICADO
const int evtLiveBlood = 0x0603; // VERIFICADO

/// Respuestas de una sola palabra observadas en el anillo.
const int ackOk = 0x00;
const int ackRejected = 0xFE;

const Map<int, String> commandNames = {
  cmdSetTime: 'SettingTime',
  cmdSettingHeartMonitor: 'SettingHeartMonitor',
  cmdSettingSpo2Monitor: 'SettingSpo2Monitor',
  cmdGetDeviceInfo: 'GetDeviceInfo',
  cmdGetNowStep: 'GetNowStep',
  cmdStartMeasurement: 'AppStartMeasurement',
  cmdHistorySport: 'Health_HistorySport',
  cmdHistorySleep: 'Health_HistorySleep',
  cmdHistoryHeart: 'Health_HistoryHeart',
  cmdHistoryBlood: 'Health_HistoryBlood',
  cmdHistorySpo2: 'Health_HistorySpo2',
  evtMeasureDone: 'MeasureComplete',
  evtStoredSleep: 'StoredSleepChunk',
  evtStoredHeart: 'StoredHeartRecord',
  evtStoredBlood: 'StoredBloodRecord',
  evtStoredSpo2: 'StoredSpo2Record',
  evtLiveHeart: 'Real_UploadHeart',
  evtLiveSpo2: 'Real_UploadBloodOxygen',
  evtLiveBlood: 'Real_UploadBlood',
};

String commandName(int code) =>
    commandNames[code] ?? 'desconocido_${code.toRadixString(16).padLeft(4, '0').toUpperCase()}';

enum MeasureType {
  heart(0x00, 'Ritmo cardíaco'),
  bloodPressure(0x01, 'Presión arterial'),
  spo2(0x02, 'Oxígeno en sangre');

  const MeasureType(this.id, this.label);
  final int id;
  final String label;

  static MeasureType? fromId(int id) {
    for (final t in MeasureType.values) {
      if (t.id == id) return t;
    }
    return null;
  }

  /// Margen antes de considerar que una medición falló, con holgura sobre los
  /// tiempos medidos en el anillo real (§6 de docs/PROTOCOLO-R21M.md).
  Duration get expectedDuration => switch (this) {
        MeasureType.heart => const Duration(seconds: 90),
        MeasureType.bloodPressure => const Duration(seconds: 90),
        MeasureType.spo2 => const Duration(seconds: 120),
      };
}

Uint8List frameOf(int code, [List<int> payload = const []]) =>
    buildFrame(code >> 8, code & 0xFF, payload);

// --- Constructores ---------------------------------------------------------

/// `01 00` + `[año u16 LE][mes][día][h][m][s][00]`.
///
/// Hay que llamarlo en cada conexión: el RTC del anillo no avanza solo y los
/// registros heredan la última hora escrita.
Uint8List setTime(DateTime when) => frameOf(cmdSetTime, [
      when.year & 0xFF,
      (when.year >> 8) & 0xFF,
      when.month,
      when.day,
      when.hour,
      when.minute,
      when.second,
      0x00,
    ]);

/// `02 00` con payload ASCII "GC".
Uint8List getDeviceInfo() => frameOf(cmdGetDeviceInfo, [0x47, 0x43]);

Uint8List getNowStep() => frameOf(cmdGetNowStep);

Uint8List startMeasurement(MeasureType kind) => frameOf(cmdStartMeasurement, [0x01, kind.id]);

Uint8List stopMeasurement() => frameOf(cmdStartMeasurement, [0x00, 0x00]);

Uint8List setHeartMonitor({required bool enabled, required int minutes}) =>
    frameOf(cmdSettingHeartMonitor, [enabled ? 1 : 0, minutes & 0xFF]);

Uint8List setSpo2Monitor({required bool enabled, required int minutes}) =>
    frameOf(cmdSettingSpo2Monitor, [enabled ? 1 : 0, minutes & 0xFF]);

Uint8List queryHistory(int code) => frameOf(code);
